src/include/storage/index/hash_index_builder.h

```diff
@@ -57,49 +56,67 @@ class HashIndexBuilder final : public InMemHashIndex, BaseHashIndex<T, S> {
     // Reserves space for at least the specified number of elements.
     void bulkReserve(uint32_t numEntries) override;
 
-    // Note: append assumes that bulkReserve has been called before it and the index has reserved
-    // enough space already.
     bool append(T key, common::offset_t value);
     bool lookup(T key, common::offset_t& result);
 
-    // Non-thread safe. This should only be called in the copyCSV and never be called in parallel.
     void flush() override;
 
 private:
-    template<bool IS_LOOKUP>
-    bool lookupOrExistsInSlotWithoutLock(Slot<S>* slot, T key, common::offset_t* result = nullptr);
-    void insertToSlotWithoutLock(Slot<S>* slot, T key, common::offset_t value);
     Slot<S>* getSlot(const SlotInfo& slotInfo);
 
-    inline Slot<S> getSlot(
-        transaction::TransactionType /*trxType*/, const SlotInfo& slotInfo) override {
-        return *getSlot(slotInfo);
-    }
-    void updateSlot(const SlotInfo& slotInfo, const Slot<S>& slot) override {
-        *getSlot(slotInfo) = slot;
-    }
     uint32_t allocatePSlots(uint32_t numSlotsToAllocate);
-    inline uint32_t appendPSlot() override { return allocatePSlots(1); }
     uint32_t allocateAOSlot();
-    uint64_t appendOverflowSlot(Slot<S>&& slot) override {
-        auto index = allocateAOSlot();
-        updateSlot(SlotInfo{index, SlotType::OVF}, slot);
-        return index;
+    /*
+     * When a slot is split, we add a new slot, which ends up with an
+     * id equal to the slot to split's ID + (1 << header.currentLevel).
+     * Values are then rehashed using a hash index which is one bit wider than before,
+     * meaning they either stay in the existing slot, or move into the new one.
+     */
+    void splitSlot(HashIndexHeader& header);
+
+    inline bool equals(
+        T keyToLookup, const S& keyInEntry, const InMemFile* /*inMemOverflowFile*/) const {
+        return keyToLookup == keyInEntry;
     }
 
-    inline bool equals(T keyToLookup, const S& keyInEntry, const InMemFile* /*inMemOverflowFile*/) {
-        return keyToLookup == keyInEntry;
+    inline void updateForNewEntry(Slot<S>* slot, uint8_t entryPos) {
+        slot->header.setEntryValid(entryPos);
+        slot->header.numEntries++;
     }
 
-    // Overridden so that we can specialize for strings
-    inline void insert(T key, uint8_t* entry, common::offset_t offset) override {
-        return BaseHashIndex<T, S>::insert(key, entry, offset);
+    inline void insert(
+        T key, Slot<S>* slot, uint8_t entryPos, common::offset_t value, InMemFile* /*memFile*/) {
+        auto entry = slot->entries[entryPos].data;
+        memcpy(entry, &key, sizeof(T));
+        memcpy(entry + sizeof(T), &value, sizeof(common::offset_t));
+        updateForNewEntry(slot, entryPos);
+    }
+    void copy(const uint8_t* oldEntry, slot_id_t newSlotId);
+    inline void insertToNewOvfSlot(
```

Comment:
This function doesn't seem likely to end as inline? Consider moving to cpp file then.